src/sentry/integrations/bitbucket_server/webhook.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import absolute_import
+
+import logging
+
+import six
+
+from datetime import datetime
+from django.db import IntegrityError, transaction
+from django.http import HttpResponse, Http404
+from django.utils import timezone
+from django.utils.decorators import method_decorator
+from django.views.decorators.csrf import csrf_exempt
+from django.views.generic.base import View
+from simplejson import JSONDecodeError
+from sentry.models import Commit, CommitAuthor, Organization, Repository
+from sentry.plugins.providers import IntegrationRepositoryProvider
+from sentry.utils import json
+
+logger = logging.getLogger("sentry.webhooks")
+
+PROVIDER_NAME = "integrations:bitbucket_server"
+
+
+class Webhook(object):
+    def __call__(self, organization, integration_id, event):
+        raise NotImplementedError
+
+    def update_repo_data(self, repo, event):
+        """
+        Given a webhook payload, update stored repo data if needed.
+        """
+
+        name_from_event = event["repository"]["project"]["key"] + "/" + event["repository"]["slug"]
+        # build the URL manually since it doesn't come back from the API in
+        # the form that we need
+        url_from_event = event["repository"]["links"]["self"][0]["href"]
+
+        if (
+            repo.name != name_from_event
+            or repo.config.get("name") != name_from_event
+            or repo.url != url_from_event
+        ):
+            repo.update(name=name_from_event, config=dict(repo.config, name=name_from_event))
+
+
+class PushEventWebhook(Webhook):
+    def __call__(self, organization, integration_id, event):
+        authors = {}
+
+        try:
+            repo = Repository.objects.get(
+                organization_id=organization.id,
+                provider=PROVIDER_NAME,
+                external_id=six.text_type(event["repository"]["id"]),
+            )
+        except Repository.DoesNotExist:
+            raise Http404()
+
+        client = repo.get_provider().get_installation(integration_id, organization.id).get_client()
+
+        # while we're here, make sure repo data is up to date
+        self.update_repo_data(repo, event)
+
+        [project_name, repo_name] = repo.name.split("/")
+
+        for change in event["changes"]:
+            for commit in client.get_commits(
+                project_name, repo_name, change.get("fromHash"), change.get("toHash")
```

Comment:
@pierredup So I am getting an error when I try to push to a repo for the first time:
```
19:55:02 server  | "POST /api/0/organizations/sentry/repos/ HTTP/1.1" 201 2431
19:55:21 server  | Traceback (most recent call last):
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/core/handlers/exception.py", line 41, in inner
19:55:21 server  |     response = get_response(request)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
19:55:21 server  |     response = self._get_response(request)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
19:55:21 server  |     response = self.process_exception_by_middleware(e, request)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
19:55:21 server  |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
19:55:21 server  |     return self.dispatch(request, *args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/utils/decorators.py", line 67, in _wrapper
19:55:21 server  |     return bound_func(*args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
19:55:21 server  |     return view_func(*args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/utils/decorators.py", line 63, in bound_func
19:55:21 server  |     return func.__get__(self, type(self))(*args2, **kwargs2)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/webhook.py", line 116, in dispatch
19:55:21 server  |     return super(BitbucketServerWebhookEndpoint, self).dispatch(request, *args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/views/generic/base.py", line 88, in dispatch
19:55:21 server  |     return handler(request, *args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/webhook.py", line 157, in post
19:55:21 server  |     handler()(organization, integration_id, event)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/webhook.py", line 68, in __call__
19:55:21 server  |     project_name, repo_name, change.get("fromHash"), change.get("toHash")
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/client.py", line 144, in get_commits
19:55:21 server  |     params={"since": fromHash, "until": toHash, "merges": "exclude"},
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/client.py", line 230, in get
19:55:21 server  |     return self.request("GET", *args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/client.py", line 224, in request
19:55:21 server  |     return self._request(*args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/client.py", line 214, in _request
19:55:21 server  |     raise ApiError.from_response(resp)
19:55:21 server  | ApiError: {"errors":[{"context":null,"message":"Commit '0000000000000000000000000000000000000000' does not exist in repository 'new repo 3'.","exceptionName":"com.atlassian.bitbucket.commit.NoSuchCommitException"}]}
```
The fix is pretty simple though:
```
            from_hash = None if change.get("fromHash") == "0" * 40 else change.get("fromHash")
            for commit in client.get_commits(
                project_name, repo_name, from_hash, change.get("toHash")
            )["values"]:
```

If you can get this fix and it works, I am ready to merge this PR!

Reply:
Thank you @scefali, I have pushed the change. I'll see if I can do a full test and make sure everything still works